src/comment_parser.rs

```diff
@@ -0,0 +1,379 @@
+//! Implements markup parsing for metamath theorem comments.
+//!
+//! The [set.mm](https://github.com/metamath/set.mm/) library contains many
+//! theorems with comments, and those comments contain markup in a special syntax
+//! implemented by [metamath.exe](https://github.com/metamath/metamath-exe).
+//! This module implements a SAX-style parser which yields events about the
+//! beginning and end of each syntax event.
+//!
+//! The [`CommentItem`] type does not contain any byte buffers directly;
+//! instead everything uses spans relative to the segment in which this comment
+//! appeared. Note that for many of the [`Span`]-containing fields the
+//! corresponding byte string in the file has to be unescaped before
+//! interpretation, using the [`CommentParser::unescape_text`] and
+//! [`CommentParser::unescape_math`] functions.
+use crate::parser::Span;
+
+/// A comment markup item, which represents either a piece of text from the input
+/// or some kind of metadata item like the start or end of an italicized group.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum CommentItem {
```

Comment:
What about the "Parentheticals"? (as they are named in the Mathemath book)
```
Contributed by...
Revised by...
Modified by...
Proof shortened by...
Proof modification is discouraged.
New usage is discouraged.
```
The last two may actually be useful outside of simply displaying the proof, e.g. for a proof assistant to select a theorem to use. So we might also to find a way to access that information without having to parse the whole comment.

Reply:
I intend to implement this in a separate PR; it does not interact with markup formatting and would be implemented as a separate "pass" over the comment content.

Reply:
Ok, sounds good.

> There is a bunch of careful ordering to get right with comment markup (like how some markup works in some nested contexts but not others), so I thought it important to have a dedicated parser to make it easier to get right.

Yes, that's better - and probably faster - to do it like that.